test/base/src/androidTest/kotlin/io/realm/kotlin/test/shared/RealmSetTests.kt

```diff
@@ -0,0 +1,1137 @@
+            assertFailsWith<IllegalStateException> {
+                set.iterator().hasNext()
+            }
+            assertFailsWith<IllegalStateException> {
+                set.iterator().next()
+            }
+            assertFailsWith<IllegalStateException> {
+                set.iterator().remove()
+            }
+        }
+    }
+
+    override fun assertContainerAndCleanup(assertion: (RealmSetContainer) -> Unit) {
+        val container = realm.query<RealmSetContainer>()
+            .first()
+            .find()
+        assertNotNull(container)
+
+        // Assert
+        errorCatcher {
+            assertion(container)
+        }
+
+        // Clean up
+        realm.writeBlocking {
+            delete(findLatest(container)!!)
+        }
+    }
+}
+
+/**
+ * Tester for generic types.
+ */
+internal class GenericSetTester<T>(
+    realm: Realm,
+    typeSafetyManager: SetTypeSafetyManager<T>,
+    private val classifier: KClassifier
+) : ManagedSetTester<T>(realm, typeSafetyManager) {
+
+    override fun toString(): String = classifier.toString()
+
+    override fun assertStructuralEquality(expected: Collection<T>, actual: Collection<T>) {
+        assertEquals(expected.size, actual.size)
+        actual.forEach {
+            assertTrue(expected.contains(it))
+        }
+    }
+
+    override fun structuralContains(collection: Collection<T>, element: T?): Boolean =
+        collection.contains(element)
+}
+
+/**
+ * Tester for ByteArray.
+ */
+internal class ByteArraySetTester(
+    realm: Realm,
+    typeSafetyManager: SetTypeSafetyManager<ByteArray>,
+    private val classifier: KClassifier
+) : ManagedSetTester<ByteArray>(realm, typeSafetyManager) {
+
+    override fun toString(): String = classifier.toString()
+
+    override fun assertStructuralEquality(
+        expected: Collection<ByteArray>,
+        actual: Collection<ByteArray>
+    ) {
+        assertEquals(expected.size, actual.size)
+
+        // We can't iterate by index on the set and the positions are not guaranteed to be the same
+        // as in the dataset so to compare the values are the same we need to bend over backwards...
+        var assertionSucceeded = 0
+        actual.forEach { actualByteArray ->
+            expected.forEach { expectedByteArray ->
+                try {
+                    assertContentEquals(expectedByteArray, actualByteArray)
+                    assertionSucceeded += 1
+                } catch (e: AssertionError) {
+                    // Do nothing, the byte arrays might be structurally equal in the next iteration
+                }
+            }
+        }
+        if (assertionSucceeded != expected.size) {
+            fail("None of the ByteArray elements were found in the dataset.")
+        }
+    }
+
+    override fun structuralContains(
+        collection: Collection<ByteArray>,
+        element: ByteArray?
+    ): Boolean {
+        var assertionSucceeded = 0
+        collection.forEach { expectedByteArray ->
```

Comment:
Again a bit convoluted. At the least add a small comment about why it can only have one match (`assertionSucceded == 1`). That would speed up reviewing and maintenance.